include/cxx.h

```diff
@@ -826,6 +826,87 @@ template <typename T>
 Box<T>::Box(uninit) noexcept {}
 #endif // CXXBRIDGE1_RUST_BOX
 
+#ifndef CXXBRIDGE1_RUST_OPTION
+template <typename T>
+class Option final {
+public:
+  Option() noexcept;
+  Option(Option&&) noexcept;
+  Option(T&&) noexcept;
+  ~Option() noexcept;
+  void drop() noexcept;
+
+  const T *operator->() const;
+  const T &operator*() const;
+  T *operator->();
+  T &operator*();
+
+  bool has_value() const noexcept;
+  const T& value() const;
+
+  union OptionContents {
+    T value;
+    size_t empty;
+
+    OptionContents() {
+      empty = 0;
+    }
+    // Destruction handled in Option since it knows if type was created
+    ~OptionContents() {}
+  };
+  OptionContents inner;  // Avoid initialization
+};
+#endif // CXXBRIDGE1_RUST_OPTION
+
+#ifndef CXXBRIDGE1_RUST_OPTION
+#define CXXBRIDGE1_RUST_OPTION
+template <typename T>
+Option<T>::Option() noexcept {}
+
+template <typename T>
+Option<T>::Option(Option&& other) noexcept {
+  if (other.inner.empty != 0) {
+    inner.value = std::move(other.inner.value);
+  }
+  other.inner.empty = 0;
+}
+
+template <typename T>
+Option<T>::~Option() noexcept {
+  this->drop();
```

Comment:
I'm not sure if I could get away with:
```
if (inner.empty != 0) {
    inner.value.~T()
}
```
(or something like that at least) or really what the implications of that are

Reply:
This bit is tricky. For `Option<T>::Option(Option&& other)`, you're missing some calls to destructors (or assuming a previously constructed object. This works if the type T is trivially constructible, but is potentially UB if we ever relax that constraint.

```
inner.value = std::move(other.inner.value);
```

This assumes that `value` was previously constructed, which is not the case as `value` is a member of a union. Use `new (&inner.value) T(std::move(other.inner.value))` instead.

```
other.inner.empty = 0;
```

This does not call the destructor of `other.inner.value`, which is potentially UB. Use `reset`/`drop` here.

If you introduce a `construct` private helper, you'll be able to use that across multiple places (e.g. if you plan to expose an `assign` member function. You might find `folly::Optional` as good inspiration here.

```
template <typename T>
Option<T>::Option(Option&& other) noexcept {
  if (other.has_value()) {
    construct(std::move(other.inner.value));
    other.reset();
  }
}

template <typename T>
Option<T>::~Option() noexcept {
  this->reset();
}

template <typename T>
Option<T>::reset() noexcept {
  this->drop();
}

template <typename... Args>
Option<T>::construct(Args&&... args) noexcept {
    new (&inner.value) T(std::forward<Args>(args)...);
    assert(innner.empty != 0);
}
```

Reply:
`Option<T>::Option(Option&& other)` is a constructor so `this` doesn't exist yet, but you are correct for (currently missing) `operator=(Option&&)`. Thanks for the placement new, i definitely dont understand that but other bindings are doing something similar